k8s/policies.go

```diff
@@ -0,0 +1,238 @@
+		deployment.Spec.Template.Spec.RuntimeClassName = p.RuntimeClassName
+	}
+
+	if p.Affinity != nil {
+		// use a replacement strategy because it is not clear that merging affinities will
+		// actually produce a meaning Affinity definition, it would likely result in
+		// an impossible to satisfy constraint
+		deployment.Spec.Template.Spec.Affinity = p.Affinity
+	}
+
+	if p.PodSecurityContext != nil {
+		if deployment.Spec.Template.Spec.SecurityContext == nil {
+			deployment.Spec.Template.Spec.SecurityContext = &corev1.PodSecurityContext{}
+		}
+
+		p.PodSecurityContext.DeepCopyInto(deployment.Spec.Template.Spec.SecurityContext)
+	}
+
+	return deployment
+}
+
+// Remove is the inverse of Apply, removing the mutations that the Policy would have applied
+func (p Policy) Remove(deployment *appsv1.Deployment) *appsv1.Deployment {
+
+	for _, policyToleration := range p.Tolerations {
+		// filter the existing tolerations and then update the deployment
+		// filter without allocation implementation from
+		// https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating
+		newTolerations := deployment.Spec.Template.Spec.Tolerations[:0]
+		for _, toleration := range deployment.Spec.Template.Spec.Tolerations {
+			if !reflect.DeepEqual(policyToleration, toleration) {
+				newTolerations = append(newTolerations, toleration)
+			}
+		}
+		deployment.Spec.Template.Spec.Tolerations = newTolerations
+	}
+
+	if p.RuntimeClassName != nil {
+		if equalStrings(deployment.Spec.Template.Spec.RuntimeClassName, p.RuntimeClassName) {
+			deployment.Spec.Template.Spec.RuntimeClassName = nil
+		}
+	}
+
+	if p.Affinity != nil && reflect.DeepEqual(p.Affinity, deployment.Spec.Template.Spec.Affinity) {
+		deployment.Spec.Template.Spec.Affinity = nil
+	}
+
+	if p.PodSecurityContext != nil {
+		sc := deployment.Spec.Template.Spec.SecurityContext
+
+		if reflect.DeepEqual(p.PodSecurityContext.SELinuxOptions, sc.SELinuxOptions) {
+			deployment.Spec.Template.Spec.SecurityContext.SELinuxOptions = nil
+		}
+		if reflect.DeepEqual(p.PodSecurityContext.SELinuxOptions, sc.WindowsOptions) {
+			deployment.Spec.Template.Spec.SecurityContext.WindowsOptions = nil
+		}
+		if p.PodSecurityContext.RunAsUser != nil {
+			deployment.Spec.Template.Spec.SecurityContext.RunAsUser = nil
+		}
+		if p.PodSecurityContext.RunAsGroup != nil {
+			deployment.Spec.Template.Spec.SecurityContext.RunAsGroup = nil
+		}
+		if p.PodSecurityContext.RunAsNonRoot != nil {
+			deployment.Spec.Template.Spec.SecurityContext.RunAsNonRoot = nil
+		}
+		if p.PodSecurityContext.SupplementalGroups != nil {
+			deployment.Spec.Template.Spec.SecurityContext.SupplementalGroups = nil
+		}
+		if p.PodSecurityContext.FSGroup != nil {
+			deployment.Spec.Template.Spec.SecurityContext.FSGroup = nil
+		}
+		if p.PodSecurityContext.Sysctls != nil {
+			deployment.Spec.Template.Spec.SecurityContext.Sysctls = nil
+		}
+	}
+
+	return deployment
+}
+
+func equalStrings(a, b *string) bool {
```

Comment:
Is there anything in the Go stdlib that can do this?

Reply:
yes, `reflect.DeepEqual` will do the same, but I didn't think anyone would want to see that in the implementation code. Some people don't like `reflect`

Reply:
do you want me to switch to reflect?